### PR TITLE
feat(cdk/a11y): Add a new InputModalityDetector service to detect the user's current input modality.

### PR DESCRIPTION
**Pull request by zelliott**
*Note: This is a first-stab at an input modality detection service and I expect there to be some churn regarding its exact behavior, so feel free to leave many comments.*

### Motivation

In our app, we conditionally render strong focus styles on elements if the element is determined to have been focused via keyboard. This is done by attaching a global `FocusMonitor` to the `document.body`. One problem with this approach is that we can only detect input modality changes when focus also changes (see https://github.com/angular/components/issues/21555 for why this is a problem with `aria-activedescendant` elements). The purpose of the `InputModalityDetector` service is to allow users to detect a broader array of input modality changes, regardless of whether focus changes.

### Approach

The `InputModalityDetector` attempts to detect input modality by listening to `keydown`, `mousedown`, and `touchstart` events, and attributing each event to its corresponding input modality (`'keyboard'`, `'mouse'`, `'touch'`). For example, when a `keydown` event is fired, the service detects the `'keyboard'` input modality. I chose this small subset of events as I thought they best represented legitimate, intentional, and committed user interaction.

The service uses some additional heuristics under-the-hood to support scenarios when these events are fired but likely don't represent an actual change in input modality. For example:

* By default, the service ignores modifier keys (please see the comment in the source for why). We allow the user to override this behavior if they wish, but would cause some difference in behavior between VoiceOver and other SRs.
* The service ignores `mousedown` events that occur too closely after a `touchstart` event (again, please see the comments in the source for why).

### Demo

https://stackblitz.com/edit/angular-ivy-9cxqav?file=src/app/input-modality-detector.ts

### Behavior

The service essentially works as you would expect when interacting with a page. Interesting edge cases are as follows:

* The input modality is `null` when the service is first instantiated, as we have no clue what the user's input modality is.
* Interacting with an inert element (e.g. clicking empty space or just pressing a random key) changes the input modality accordingly. My current opinion is that this is fine, as it represents an *attempted interaction* with the page.
* If a user's input modality changes in the middle of interacting with a built-in popup control such as a `<select>` or `<input type="datepicker">`, this change is not detected. This is because no events are emitted.
* If a user clicks while pressing a modifier key such as Control or Option (to either "right-click" or "open in a new tab"), this is detected as mouse input modality (as modifier keys are ignored by default, as mentioned earlier).
* If a user is using a screen reader, input modality *for the most part* will not be detected (there's a comment in the source that goes into more detail).

### Testing

I'm planning on testing the service on the following platforms, browsers, and screen readers:

* Platforms: Web (Windows, Mac, Chrome OS), Mobile (Android, iOS)
* Browsers: Chrome, FF, Safari, Edge, IE11.
* SRs: NVDA, JAWS, VoiceOver, ChromeVox, Android Accessibility Suite.

I've currently tested *most* of the combinations above, with the exception of Chrome OS (+ChromeVox), Android (+Android Accessibility Suite), and IE11.

### Prior Art / Resources

* https://github.com/ten1seven/what-input
* https://github.com/paflakst/office-ui-fabric-react/blob/master/packages/utilities/src/initializeFocusRects.ts
* https://github.com/medialize/ally.js/blob/master/src/observe/interaction-type.js
* https://github.com/Modernizr/Modernizr/issues/869
* https://github.com/rafgraph/event-from

### CODE REVIEW

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [inline src/cdk/a11y/input-modality/input-modality-detector.ts:84]: Note to self that I still need to evaluate to what extent we can replace `FocusMonitor`'s own input modality detection with this service's.

**Review by jelbourn — COMMENTED** [review]
Overall this looks good; if we can show that `FocusMonitor` can use this it would be good to go IMO

css @mmalerba and @crisbeto who have done most of the work on `FocusMonitor`

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (crisbeto) [inline src/cdk/a11y/input-modality/input-modality-detector.ts:50]: These are all modifier keys so shouldn't we be looking at `KeyboardEvent.altKey` etc.? We have some helpers for dealing with modifiers in `@angular/cdk/keycodes` already.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (crisbeto) [inline src/cdk/a11y/input-modality/input-modality-detector.ts:125]: nit: IMO the early `return` here and in `_onTouchStart` makes things harder to follow, because there's so little logic after it. It might be clearer to just wrap the rest of the logic in an `if`.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (crisbeto) [inline src/cdk/a11y/input-modality/input-modality-detector.ts:94]: You could replace the `ReplaySubject` with a `BehaviorSubject` which has identical behavior, but it also allows you to read the latest value which can then be used for the `inputModality` getter.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (crisbeto) [inline src/cdk/a11y/input-modality/input-modality-detector.ts:168]: All of these events should be bound outside of the Angular zone.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (crisbeto) [inline src/cdk/a11y/input-modality/input-modality-detector.ts:165]: nit: extra space after the equals sign.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (crisbeto) [inline src/cdk/a11y/input-modality/input-modality-detector.ts:151]: AFAIK there are some issues in ViewEngine with typing constructor parameters as `Document`. The way we work around it in other places is to type the parameter as `any` and then cast it to `Document` afterwards.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (crisbeto) [inline src/cdk/a11y/input-modality/input-modality-detector.ts:170]: I'm still a little iffy about maintaining several listeners on the `document`, not matter whether anything is actually using them. If we were to remove the `inputModality` getter and only keep an observable as the way of knowing what the current modality is, we could add/remove the listeners depending on the number of subscribers.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (crisbeto) [inline src/cdk/a11y/input-modality/input-modality-detector.ts:174]: This will throw in a non-browser environment.

**Comment by zelliott** [comment]
> if we can show that `FocusMonitor` can use this it would be good to go IMO

I think I responded to all open comments, but I still need to evaluate integration with `FocusMonitor`.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts]: Whoops, thanks. Fixed.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts]: Good point. I'll defer to Jeremy & Miles on this. The downside is, of course, there's no easy way to just get the most recently detected input modality.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts:156]: I did notice this when looking at other services. Do you recall how to reproduce these issues? I'd prefer to be able to reproduce the issue to confirm the `any` + cast is actually needed, but if you're confident, then I'll just go ahead and make the change.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts]: Done.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts]: Thanks, I made this change, but it's causing some problems in the dev-app demo and Stackblitz. If you take a look at the Stackblitz (I updated with the new implementation of binding these event listeners outside of the ngZone), you'll notice that the input modality no longer updates in the view (though of course the subscription still emits). Without knowing too much about zones my it seems like Angular isn't running CD because it doesn't think anything is changing. Is this expected behavior? If so, I'll need to update the dev-app demo accordingly.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts:98]: This was actually my previous implementation, but I ended switching to a `ReplaySubject` because I didn't like how the `BehaviorSubject` emitted `null` at the start. My thinking was that `inputModalityChange` should only emit when the input modality changes, and I didn't want subscribers to have to filter out the first `null` value in their logic. WDYT?

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts:129]: I almost always prefer returning early to reduce nesting. I'll leave this as-is for now, but if Jeremy/Miles/anyone has strong opinions either way, I'll switch to your recommendation.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts]: I considered that, but I wanted to give users the ability to configure which keys they'd like to ignore/not ignore, and it seemed easiest to support that via a single `ignoreKeys` API.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts:100]: Done.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts:95]: Only to prevent people from setting it themselves for whatever reason. Happy to change it to a public property, just LMK.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts]: Done.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts]: Gotcha. Switched to `keyCode`.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts]: Done.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (crisbeto) [reply src/cdk/a11y/input-modality/input-modality-detector.ts]: It could have a `modifiers` array and `ignoreKeys`. If I remember correctly, the keydown will dispatch `Alt` if you just press the `Alt` key once, but if you pressed `Alt` + `DOWN_ARROW`, you'd get a `DOWN_ARROW` event with the `Alt` modifier.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (crisbeto) [reply src/cdk/a11y/input-modality/input-modality-detector.ts:98]: You could add `.pipe(skip(1))` when exposing the subject publicly. It's not a bit deal to leave it as it is though.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (crisbeto) [reply src/cdk/a11y/input-modality/input-modality-detector.ts:156]: I think that we were running into it in our Angular Universal integration tests. There's a readme under `src\universal-app\DEBUG.md`.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (crisbeto) [reply src/cdk/a11y/input-modality/input-modality-detector.ts]: If the listener is bound outside of the zone, you'll have to use `NgZone.run` when updating the view, otherwise change detection won't run. It's a little cumbersome, but we may end up triggering unnecessary change detections otherwise.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts]: Note that now the getter is more necessary because we're grabbing the value from the underlying `BehaviorSubject`.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts]: From my testing, if I press `Alt` + `Down Arrow`, I get both an `Alt` and a `Down Arrow` event, so this should work fine (I tested with a number of modifier keys and on both Mac and Windows). Changing the API to both `modifiers` and `ignoreKeys` increases its surface error, and it's not clear to me what the benefit is (do we think it'll be clearer to users?).

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts]: Okay, I don't feel strongly so I'll switch to the `BehaviorSubject`.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts]: Gotcha, done.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts]: I added the `InputModalityDetector` to the universal kitchen sink app as-is, and ran `bazel test //src/universal-app:server_test --test_output=all`. The target passed, and I didn't see any relevant console warnings/logs. Is there something in particular I should be looking for?

**Comment by zelliott** [comment]
It's possible to update `FocusMonitor` to use this new `InputModalityDetector` service to determine focus origin, but it requires us to tweak the API of this service a bit, and causes some changes in `FocusMonitor`'s behavior. https://github.com/angular/components/commit/db685c0f54c7ac788eecb3fafba6e6392e910ee0 demonstrates one potential integration. I left comments on the commit above documenting various things and explaining any changes in behavior.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (mmalerba) [inline src/cdk/a11y/input-modality/input-modality-detector.ts:61]: should this be in `InputModalityDetectorOptions`?

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (mmalerba) [reply src/cdk/a11y/input-modality/input-modality-detector.ts]: I think in most cases the people will want the simple check the last `inputModality` API. The listeners don't do much and are marked as `passive` so I don't think it'll be an issue

**Review by mmalerba — APPROVED** [review]
LGTM

@jelbourn do you want to put this straight into master or create a feature branch so that the focus monitor can be migrated to use this before merging to master?

**Review by mmalerba — APPROVED** [review]
LGTM

@jelbourn do you want to put this straight into master or create a feature branch so that the focus monitor can be migrated to use this before merging to master?

**Comment by zelliott** [comment]
> @jelbourn do you want to put this straight into master or create a feature branch so that the focus monitor can be migrated to use this before merging to master?

I'd recommend a feature branch. Please see https://github.com/angular/components/pull/22371#issuecomment-814277510 regarding integration with `FocusMonitor`. I left a number of open comments on the commit linked to in that comment.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts:61]: Hm, it wasn't configurable in `FocusMonitor` and I'm not sure I see a need for different users wanting different values here. If a user is configuring this, chances are our value is simply wrong and should be updated?

**Comment by jelbourn** [comment]
I've updated the PR to target a feature branch called `input-modality`

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (jelbourn) [inline src/cdk/a11y/input-modality/input-modality-detector.ts:153]:
```suggestion
      @Inject(DOCUMENT) document: Document|Document,
```
Using `Document` here will break in SSR because TypeScript preserves type information on decorated params, and `Document` doesn't existing in a nodejs environment. We can use this weird workaround to trick TS into not emitting the type as-is. 

Can go in a follow-up, but you can test this by adding something to make this API run as part of our `kitchen-sink.ts` component (which is our SSR test component).

Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts:153]: See the thread at https://github.com/angular/components/pull/22371#discussion_r605479216. The kitchen sink e2e test seems to pass - what should I be looking for?

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts]: Done.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts]: I don't feel strongly, changed to `modalityChanges` to align with `mostRecentModality`.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts]: Done.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.ts:88]: See https://github.com/angular/components/commit/db685c0f54c7ac788eecb3fafba6e6392e910ee0

**Comment by zelliott** [comment]
@jelbourn I think I may have accidentally pulled in a commit from annieyw (https://github.com/angular/components/pull/22371/commits/212c02655e888aea3f453f12a99df1703c48f874) when rebasing off of master. LMK if there's something I should do on my end.

**Comment by mmalerba** [comment]
I synced input-modality up to master, it should go away if you rebase on input-modality

**Comment by zelliott** [comment]
Tried that (and also tried an empty commit to get the rebase picked up), to no avail. It's weird because comparing local with remote here (https://github.com/angular/components/compare/input-modality...zelliott:input-modality?expand=1) shows the commits I'd expect, whereas this PR has an extra one.

**Comment by zelliott** [comment]
Ok fixed by squashing.

File: src/cdk/a11y/input-modality/input-modality-detector.spec.ts
Comment (raymondji) [inline src/cdk/a11y/input-modality/input-modality-detector.spec.ts:146]: nit: maybe remove "additional"? It doesn't seem like there is a base set of keys that are ignored, if `ignoreKeys` is passed in.

Comment (zelliott) [reply src/cdk/a11y/input-modality/input-modality-detector.spec.ts]: Done.

File: src/cdk/a11y/input-modality/input-modality-detector.ts
Comment (jelbourn) [reply src/cdk/a11y/input-modality/input-modality-detector.ts:153]: Hmm, maybe something changed in the tooling so that we don't have to worry about this any more ¯\\_(ツ)_/¯

**Review by jelbourn — APPROVED** [review]
lgtm - I also fixed the rebase issue

**Pull request merged**
